core/lib/zksync_core/src/state_keeper/metrics.rs

```diff
@@ -395,3 +399,63 @@ pub(super) struct ExecutorMetrics {
 
 #[vise::register]
 pub(super) static EXECUTOR_METRICS: vise::Global<ExecutorMetrics> = vise::Global::new();
+
+#[derive(Debug, Metrics)]
+#[metrics(prefix = "block_tip")]
+pub(crate) struct BlockTipMetrics {
+    #[metrics(buckets = Buckets::exponential(60000.0..=80000000.0, 2.0))]
+    gas_used: Histogram<usize>,
+    #[metrics(buckets = Buckets::exponential(1.0..=60000.0, 2.0))]
+    pubdata_published: Histogram<usize>,
+    #[metrics(buckets = Buckets::exponential(1.0..=4096.0, 2.0))]
+    circuit_statistic: Histogram<usize>,
+    #[metrics(buckets = Buckets::exponential(1.0..=4096.0, 2.0))]
+    execution_metrics_size: Histogram<usize>,
+    #[metrics(buckets = Buckets::exponential(1.0..=60000.0, 2.0))]
+    block_writes_metrics_size: Histogram<usize>,
+    #[metrics(buckets = Buckets::exponential(1.0..=60000.0, 2.0))]
+    block_writes_metrics_gas_count: Histogram<usize>,
+}
+
+impl BlockTipMetrics {
+    pub fn observe(&self, execution_result: &VmExecutionResultAndLogs) {
+        self.gas_used
+            .observe(execution_result.statistics.gas_used as usize);
+        self.pubdata_published
+            .observe(execution_result.statistics.pubdata_published as usize);
+        self.circuit_statistic
+            .observe(execution_result.statistics.circuit_statistic.total());
+        self.execution_metrics_size
+            .observe(execution_result.get_execution_metrics(None).size());
+
+        let initial_writes = execution_result
+            .logs
+            .storage_logs
+            .iter()
+            .filter(|log| log.log_type == StorageLogQueryType::InitialWrite)
+            .count();
+        let repeated_writes = execution_result
+            .logs
+            .storage_logs
+            .iter()
+            .filter(|log| log.log_type == StorageLogQueryType::RepeatedWrite)
+            .count();
+        let total_updated_values_size = execution_result.logs.storage_logs.len() * (32 + 1); // 32 bytes per value + 1 metadata byte
+
+        let writes_metrics = DeduplicatedWritesMetrics {
+            initial_storage_writes: initial_writes,
+            repeated_storage_writes: repeated_writes,
+            total_updated_values_size,
+        };
```

Comment:
I think in the current form it doesn't make much sense, without taking into deduplication and compression values we get will be very inaccurate. What I would do instead is to measure the difference between [these](https://github.com/matter-labs/zksync-era/blob/main/core/lib/zksync_core/src/state_keeper/keeper.rs#L694)
metrics and write metrics before batch tip was applied, i.e. code will look like
```
let block_writes_metrics = updates_manager
                    .storage_writes_deduplicator
                    .apply_and_rollback(logs_to_apply_iter.clone());
let diff = block_writes_metrics - updates_manager.storage_writes_deduplicator.metrics();
```
There is a caveat, I suspect diff would be negative, so just using "-" might not work, instead you can cast every metric field to signed type and subtract then. Diff may be negative because in batch tip some storage slots are reset to zero, so writes to them will be deduplicated

Reply:
How big the difference between values could be? Can it be that big, that if we use raw approach it could break something? Or we can just say that it has some overhead(in case metrics won’t report the biggest values)

Reply:
Your current approach measures an upper bound for state_diffs size published. But this upper bound is very inaccurate. For example, let's simplify batch tip execution and assume the only action it does is the following: there is a storage slot which is always written when VM executes any transaction, and at the end of the batch at "batch tip" this slot value is set back to 0. Your approach will return that during batch tip there was 1 initial storage write, 0 repeated storage writes, total_updated_values_size = 33. But in fact state diffs size didn't increase by `(length of 1 initial write + total_updated_values_size)`, instead it decreased by that value.
Note that an example with storage slot set to 0 is not far from reality, in the actual batch tip execution there are several such slots that are rolled back.

Reply:
Okay, I got your point, I don’t think that negative diff really matters to us since we only need to know upper bound, so we probably can just set 0 values where it really doesn’t matter for us